### PR TITLE
[sc-143215] Adding an "append to list" recipe

### PR DESCRIPTION
**Pull request by alexbourret**

### CODE REVIEW

**Review by liamlynch-data — COMMENTED** [review]
I've tested the functionality out and it  works fine. 
(Specifically, when using the Append Recipe to either write or append, it all works and the id in the list URLs you get by choosing "Share" do not change and they still link back to the same (updated) list, and nothing is moved to the recycle bin.)

I have a few of questions on the functionality choices made though - just to understand them really.

1. Why is it called append when it is not necessarily an append? (there is a setting to determine if it is an append, and that also exists in the dataset version when doing a sync). Maybe there is a convention here, but "write to list" might be a better name? 

2. Why do we need an output to the append recipe (why is it mandatory) - is it a limitation of recipes? I guess the integrity of the flow might be compromised if this does not exist, is that it? (i.e. if there is downstream data, you still need to be able to run the whole flow). But if the list is just an output and we don't use the data downstream, why force people to have one? 

3. When you first set things up to output at dataset to sharepoint, for the recipe you need to create the list in sharepoint, it needs to at least exist already. You don't need to do this when you use a dataset - it will create the list for you if not there. This may be confusing to people and it would be simpler if they didn't need to do it - is there anything we could do about this? Or are there good reasons not to do anything here?

I will take a closer look at the code now.

File: custom-recipes/sharepoint-online-append-list/recipe.py [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+from sharepoint_client import SharePointClient
+
+
+logger = SafeLogger("sharepoint-online plugin", DSSConstants.SECRET_PARAMETERS_KEYS)
+logger.info('SharePoint Online append to list recipe v{}'.format(DSSConstants.PLUGIN_VERSION))
+
+
+def convert_date_format(json_row):
+    #  Convert pandas timestamps to iso
+    return json_row
```
Comment (liamlynch-data) [inline custom-recipes/sharepoint-online-append-list/recipe.py:14]: Nothing seems to be happening here

File: custom-recipes/sharepoint-online-append-list/recipe.py [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+    max_workers = config.get("max_workers", 1)
+    batch_size = config.get("batch_size", 100)
+    sharepoint_list_view_title = config.get("sharepoint_list_view_title", "")
+logger.info("init:advanced_parameters={}, max_workers={}, batch_size={}".format(advanced_parameters, max_workers, batch_size))
+metadata_to_retrieve.append("Title")
+display_metadata = len(metadata_to_retrieve) > 0
+client = SharePointClient(config)
+sharepoint_list_view_id = None
+if sharepoint_list_view_title:
+    sharepoint_list_view_id = client.get_view_id(sharepoint_list_title, sharepoint_list_view_title)
```
Comment (liamlynch-data) [inline custom-recipes/sharepoint-online-append-list/recipe.py:54]: I don't think sharepoint_list_view_id is used anywhere - and probably isn't needed when writing data?

File: custom-recipes/sharepoint-online-append-list/recipe.json [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+            "description": "sites/site_name/subsite...",
+            "visibilityCondition": "model.advanced_parameters == true"
+        },
+        {
+            "name": "sharepoint_list_view_title",
+            "label": "View name",
+            "description": "Read data from a specific view",
+            "type": "STRING",
+            "defaultValue": "",
+            "visibilityCondition": "model.advanced_parameters == true"
```
Comment (liamlynch-data) [inline custom-recipes/sharepoint-online-append-list/recipe.json:136]: Do we need this param? If we are writing data does this do anything?

**Comment by alexbourret** [comment]
Hi @liamlynch-data !
Here are some answers:
1. This recipe is restricted to append mode. The write_mode selector in `recipe.json` is hidden ignored in the recipe itself, where the value is forced to `append`. The option is indeed available in the sync recipe, but until now it is not passed to plugins. (it actually raise an error if used)
2. Output is mandatory to recipe. The concept of the flow is more about building datasets rather than running the recipes that produce these datasets.
3. It could be done, we'll discuss it

File: tests/python/integration/test_scenario.py [hunk 4]
```diff
@@ -30,5 +30,10 @@ def test_run_sharepoint_online_site_root_overwrite(user_dss_clients):
 def test_run_sharepoint_online_write_on_empty_root_path(user_dss_clients):
     dss_scenario.run(user_dss_clients, project_key=TEST_PROJECT_KEY, scenario_id="WRITEONEMPTYROOTPATH")
 
+
 def test_run_sharepoint_online_file_overwrite(user_dss_clients):
     dss_scenario.run(user_dss_clients, project_key=TEST_PROJECT_KEY, scenario_id="FILEOVERWRITE")
+
+
+def test_run_sharepoint_online_append_to_list_recipe(user_dss_clients):
```
Comment (liamlynch-data) [inline tests/python/integration/test_scenario.py:38]: I haven't been able to look at this scenario - can you provide the details or share what it does?

**Review by liamlynch-data — CHANGES_REQUESTED** [review]
The code looks basically fine to me, just a couple of things that may have been missed or need to be removed in the recipe version I think

File: parameter-sets/oauth-login/parameter-set.json [hunk 5]
```diff
@@ -5,7 +5,7 @@
         "icon": "icon-cloud"
     },
     "defaultDefinableInline": false,
-    "defaultDefinableAtProjectLevel": true,
+    "defaultDefinableAtProjectLevel": false,
```
[the remark this answers is not in the document]
Comment (alexbourret) [reply parameter-sets/oauth-login/parameter-set.json:8]: It is not related to the card but has to be done: OAuth presets do not work when defined at project level so leaving this default option set to true is confusing to the final users.

File: python-connectors/sharepoint-online_lists/connector.py [hunk 6]
```diff
@@ -44,65 +38,17 @@ def __init__(self, config, plugin_config):
...
-                    self.column_entity_property_name[column[SharePointConstants.STATIC_NAME]] = column[SharePointConstants.ENTITY_PROPERTY_NAME]
-                    self.dss_column_name[column[SharePointConstants.STATIC_NAME]] = column[SharePointConstants.TITLE_COLUMN]
-                    self.dss_column_name[column[SharePointConstants.ENTITY_PROPERTY_NAME]] = column[SharePointConstants.TITLE_COLUMN]
-                if sharepoint_type == "date":
-                    self.columns_to_format.append((column[SharePointConstants.STATIC_NAME], sharepoint_type))
-        logger.info("get_read_schema: Schema updated with {}".format(dss_columns))
-        return {
-            SharePointConstants.COLUMNS: dss_columns
-        }
+        return self.client.get_read_schema(display_metadata=self.display_metadata, metadata_to_retrieve=self.metadata_to_retrieve)
```
[the remark this answers is not in the document]
Comment (alexbourret) [reply python-connectors/sharepoint-online_lists/connector.py:44]: The get_read_schema method can be called by DSS so I can't remove this section

**Comment by liamlynch-data** [comment]
> Hi @liamlynch-data ! Here are some answers:
> 
> 1. This recipe is restricted to append mode. The write_mode selector in `recipe.json` is hidden ignored in the recipe itself, where the value is forced to `append`. The option is indeed available in the sync recipe, but until now it is not passed to plugins. (it actually raise an error if used)
> 2. Output is mandatory to recipe. The concept of the flow is more about building datasets rather than running the recipes that produce these datasets.
> 3. It could be done, we'll discuss it

Re 1 - thanks - I see now I mis-understood the input/output config, where this is selecting it for the DSS dataset output not for the sharepoint one. 
2. I see, I thought it might be something like that.
3. Ok - not a blocker for this PR though, IMO

File: parameter-sets/oauth-login/parameter-set.json (hunk 5 above)
[the remark this answers is not in the document]
Comment (liamlynch-data) [reply parameter-sets/oauth-login/parameter-set.json:8]: Fair enough!

File: custom-recipes/sharepoint-online-append-list/recipe.json (hunk 3 above)
Comment (alexbourret) [reply custom-recipes/sharepoint-online-append-list/recipe.json]: 👍 removed in [c4fabcf](https://github.com/dataiku/dss-plugin-sharepoint-online/pull/47/commits/c4fabcf3486c0379c3dec03cd8a2f093d0e04075)

File: custom-recipes/sharepoint-online-append-list/recipe.py (hunk 2 above)
Comment (alexbourret) [reply custom-recipes/sharepoint-online-append-list/recipe.py]: removed in [c4fabcf](https://github.com/dataiku/dss-plugin-sharepoint-online/pull/47/commits/c4fabcf3486c0379c3dec03cd8a2f093d0e04075)

File: tests/python/integration/test_scenario.py (hunk 4 above)
Comment (alexbourret) [reply tests/python/integration/test_scenario.py:38]: It creates a sharepoint list, appends to it and check the resulting number of rows.

File: custom-recipes/sharepoint-online-append-list/recipe.py (hunk 1 above)
Comment (alexbourret) [reply custom-recipes/sharepoint-online-append-list/recipe.py]: Fixed in [13dc852](https://github.com/dataiku/dss-plugin-sharepoint-online/pull/47/commits/13dc852d214b9625a88b222fd20ebe6255dde56e)

**Comment by liamlynch-data** [comment]
@alexbourret - have you addressed everything you want to now? Shall I re-review?

**Review by liamlynch-data — APPROVED** [review]
Changes look good and I tested with some dates that conversions/handling seems to working fine. Thanks!

I wouldn't mind those 2 remaining minor things being addressed/answered but it's not a blocker to the card/merge.

File: custom-recipes/sharepoint-online-append-list/recipe.py [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+def convert_date_format(json_row):
+    #  Convert pandas timestamps to iso
+    return json_row
+
+
+input_dataset = get_input_names_for_role('input_dataset')
+input_parameters_dataset = dataiku.Dataset(input_dataset[0])
+input_parameters_dataframe = input_parameters_dataset.get_dataframe()
+input_schema = input_parameters_dataset.read_schema()
+output_names_stats = get_output_names_for_role('api_output')
```
[the remark this answers is not in the document]
Comment (alexbourret) [reply custom-recipes/sharepoint-online-append-list/recipe.py]: Fixed in [04b9f7e](https://github.com/dataiku/dss-plugin-sharepoint-online/pull/47/commits/04b9f7ec8a5025f9811ad6b48234c12c031f19c9)

**Comment by liamlynch-data** [comment]
@alexbourret that's for those last couple of changes - all looks good, and I did a quick retest and all was still working fine for me.

**Pull request merged**